coderd/workspaceusage/tracker.go

```diff
@@ -0,0 +1,221 @@
+func (wut *Tracker) Add(workspaceID uuid.UUID) {
+	wut.m.Add(workspaceID)
+}
+
+// flush updates last_used_at of all current workspace IDs.
+// If this is held while a previous flush is in progress, it will
+// deadlock until the previous flush has completed.
+func (wut *Tracker) flush(now time.Time) {
+	// Copy our current set of IDs
+	ids := wut.m.UniqueAndClear()
+	count := len(ids)
+	if wut.flushCh != nil { // only used for testing
+		defer func() {
+			wut.flushCh <- count
+		}()
+	}
+	if count == 0 {
+		wut.log.Debug(context.Background(), "nothing to flush")
+		return
+	}
+
+	// Set a short-ish timeout for this. We don't want to hang forever.
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	// nolint: gocritic // system function
+	authCtx := dbauthz.AsSystemRestricted(ctx)
+	wut.flushLock.Lock()
+	defer wut.flushLock.Unlock()
+	if err := wut.s.BatchUpdateWorkspaceLastUsedAt(authCtx, database.BatchUpdateWorkspaceLastUsedAtParams{
+		LastUsedAt: now,
+		IDs:        ids,
+	}); err != nil {
+		wut.log.Error(ctx, "failed updating workspaces last_used_at", slog.F("count", count), slog.Error(err))
+		return
+	}
+	wut.log.Info(ctx, "updated workspaces last_used_at", slog.F("count", count), slog.F("now", now))
+}
+
+// Loop periodically flushes every tick.
+// If Loop is called after Close, it will panic. Don't do this.
+func (wut *Tracker) Loop() {
+	// Calling Loop after Close() is an error.
+	select {
+	case <-wut.doneCh:
+		panic("developer error: Loop called after Close")
+	default:
+	}
+	defer func() {
+		wut.log.Debug(context.Background(), "workspace usage tracker loop exited")
+	}()
+	for {
+		select {
+		case <-wut.stopCh:
+			close(wut.doneCh)
+			return
+		case now, ok := <-wut.tickCh:
+			if !ok {
+				return
+			}
+			wut.flush(now.UTC())
+		}
+	}
+}
+
+// Close stops Tracker and returns once Loop has exited.
+// After calling Close(), Loop must not be called.
+func (wut *Tracker) Close() {
+	wut.stopOnce.Do(func() {
+		wut.stopCh <- struct{}{}
+		wut.stopTick()
+		<-wut.doneCh
+	})
+}
+
+// uuidSet is a set of UUIDs. Safe for concurrent usage.
+// The zero value can be used.
+type uuidSet struct {
+	l sync.Mutex
+	m map[uuid.UUID]struct{}
+}
+
+func (s *uuidSet) Add(id uuid.UUID) {
+	s.l.Lock()
+	defer s.l.Unlock()
+	if s.m == nil {
+		s.m = make(map[uuid.UUID]struct{})
+	}
+	s.m[id] = struct{}{}
+}
+
+// UniqueAndClear returns the unique set of entries in s and
+// resets the internal map.
+func (s *uuidSet) UniqueAndClear() []uuid.UUID {
+	s.l.Lock()
+	defer s.l.Unlock()
+	if s.m == nil {
+		s.m = make(map[uuid.UUID]struct{})
+	}
+	l := make([]uuid.UUID, 0)
+	for k := range s.m {
+		l = append(l, k)
+	}
+	// For ease of testing, sort the IDs lexically
+	sort.Slice(l, func(i, j int) bool {
+		// For some unfathomable reason, byte arrays are not comparable?
+		return bytes.Compare(l[i][:], l[j][:]) < 0
+	})
+	s.m = make(map[uuid.UUID]struct{})
```

Comment:
I believe the idiomatic approach as of Go 1.21 is:

```suggestion
	clear(s.m)
```

https://tip.golang.org/ref/spec#Clear

Reply:
Ooh nice, forgot about `clear()`

Reply:
Oh yeah! `clear` is newer iirc. 

1.21... newish lol